### ISSUE TITLE
argsIgnorePattern exception shadows other arguments that should be flagged

### ISSUE THREAD

**Issue by machajew** [eslint/eslint#9899]
**Tell us about your environment**

* **ESLint Version:** v4.15.0
* **Node Version:** v8.9.3
* **npm Version:** v5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

module.exports = {
    "extends": "airbnb-base",
    "env": {
        "browser": true,
        "node": true
    },
    "rules": {
        "no-unused-vars": ["error", { "argsIgnorePattern": "next" }]
    }
};

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js

const app = null;

// middleware 1
app.use((req1, next) => { // the 'req1' arg is (correctly) flagged for being an unused arg
  // per my config, 'next' arg is (correctly) ignored, even though it is unused
});

// middleware 2
app.use((req2, next) => { // req2 should be flagged for being an unused arg here!
  const err = null;
  next(err); // now that I happen to use 'next' here, req2 is not being flagged
  // whether or not I use 'next' should NOT influence whether 'req2' is flagged for being unused
});



```

```bash

eslint scratch.js

```

**What did you expect to happen?**

Any function argument matching my 'argsIgnorePattern' option (i.e. the argument 'next') should not be flagged, regardless of whether it is unused (this behaved correctly).

But this should not influence whether or not any other function arguments are flagged as unused.

**What actually happened? Please include the actual, raw output from ESLint.**
The use of an argument matching my 'argsIgnorePattern' option "shadowed" another function arg that SHOULD HAVE been flagged as unused.

Output:

 4:10  error  'req1' is defined but never used. Allowed unused args must match /next/  no-unused-vars

✖ 1 problem (1 error, 0 warnings)

**Comment by machajew** [comment]
I realized this is an issue regardless of the optional argsIgnorePattern.

See this simpler code for an example of the issue:

```
const app = null;

app.use((foo) => { // foo is flagged as unused (correct)
  //
});

app.use((foo, bar) => { // foo is not flagged as unused, even though it isn't used
  bar();
});
```

**Comment by not-an-aardvark** [comment]
Is this the same issue as https://github.com/eslint/eslint/issues/9750?

**Comment by platinumazure** [comment]
Hi @machajew, thanks for the issue.

I just want to make sure: Do you still see issues if you use the [`args: "all"` option](https://eslint.org/docs/rules/no-unused-vars#args)?

There might be a few different things going on here:

* Using `args: "after-used"` (the default option for arguments) will definitely not flag arguments that occur before the latest used argument (this is by design and the remedy is to switch to `args: "all"`
* Whenever using `args: "after-used"`, only the latest unused argument is flagged, instead of all arguments that are unused (after the latest used argument). (#9750)

**Comment by machajew** [comment]
@platinumazure I apologize - in my case, this was simply a matter of using the `args: "all"` option, which I didn't realize was available/not default functionality.

Sorry about that!
